site/docs/expressions/scalar_functions.md

```diff
@@ -1,52 +1,118 @@
+
+* Integer: 64 bit signed integer (can be a literal or a parameter value)
+* Boolean: True and False
+* Type: A Substrait type (with possibly additional embedded expressions)
+
+These types are evaluated using a small set of operations to support common scenarios. List of valid operations:
+
+```
+Math: +, -, *, /, min, max
+Boolean: &&, ||, !, <, >, ==
+Parameters: type, integer
+Literals: type, integer
+```
+
+Fully defined with argument types:
+
+* `type_parameter(string name) => type`
+* `integer_parameter(string name) => integer`
+* `not(boolean x) => boolean` 
+* `and(boolean a, boolean b) => boolean` 
+* `or(boolean a, boolean b) => boolean` 
+* `multiply(integer a, integer b) => integer`
+* `divide(integer a, integer b) => integer`
+* `add(integer a, integer b) => integer`
+* `subtract(integer a, integer b) => integer`
+* `min(integer a, integer b) => integer`
+* `max(integer a, integer b) => integer`
+* `equal(integer a, integer b) => boolean`
+* `greater_than(integer a, integer b) => boolean`
+* `less_than(integer a, integer b) => boolean`
+* `covers(Type a, Type b) => boolean`Covers means that type b matches type A for as much as type B is defined. For example, if type A is `VARCHAR(20)` and type B is `VARCHAR(N)`, type B would be considered covering. Similarlily if type A was `List<Struct<a:f32, b:f32>>`and type B was `List<Struct<>>`, it would be considered covering. Note that this is directional "as in B covers A" or "B can be further enhanced to match the definition A". 
+* `if(boolean a) then (integer) else (integer)`
+* `if(boolean a) then (type) else (type)`
+
+#### Example Type Expressions Uses
+
+For reference, here are are some common output type derivations and how they can be expressed with a return type expression:
+
+| Operation                                                 | Definition                                                   |
+| --------------------------------------------------------- | ------------------------------------------------------------ |
+| Add item to list                                          | `add(<List<T>, T>) => List<T>`                               |
+| Decimal Division                                          | `divide(Decimal(P1,S1), Decimal(P2,S2)) => Decimal(P1 -S1 + S2 + MAX(6, S1 + P2 + 1), MAX(6, S1 + P2 + 1))` |
+| Do regex on only string maps to return values.            | `extract_values(Map<K,V>) => List<V> WHERE K IN [STRING, VARCHAR(N), FIXEDCHAR(N)]` |
+| Concatenate two fixed sized character strings             | `concat(FIXEDCHAR(A), FIXEDCHAR(B)) => CHAR(A+B)`            |
+| Make a struct of a set of fields and a struct definition. | `make_struct(<type> T, K...) => T`                           |
```

Comment:
Where are the field names expressed?

Reply:
In the type T

Reply:
What is the meaning of `K...` here? Does it mean every `K[0]` up to `K[n - 1]` are possibly distinct types? Or does it mean that every argument must be the same type `K`?

Also, what is the `<type> T` syntax? Isn't that covered by declaring type parameters in the function signature like `make_struct<T, K>(T, K...)`?

Reply:
>What is the meaning of K... here?
I called this out elsewhere in the doc. In a variadic case, the last argument declaration can be either marked as consistent or inconsistent (I need to add this to the protobuf). If consistent, must resolve the same. If inconsistent, each instance can be a different value.

> Also, what is the <type> T syntax?

It's all pseudo syntax to help explain the concepts for the spec. We need to define an actual syntax.